File: praktikum1/quicksort.py
def partition(a, begin, end):

    pivot = a[end]
    i = begin - 1

    for j in range (begin, end):
        if a[j] <= pivot:
            i = i + 1
            temp = a[i]
            a[i] = a[j]
            a[j] = temp

    temp = a[i+1]
    a[i+1] = a[end]
    a[end] = temp

    return i+1

def sortiere(a, begin, end):

    if begin < end:
        pivotIndex = partition(a, begin, end)
        sortiere(a, begin, pivotIndex -1)
        sortiere(a, pivotIndex +1, end)



a = [99, 0, 33, 27, 87, 80, 65, 77, 39, 88]

print(a)

sortiere(a, 0, len(a)-1)

print(a)

File: praktikum1/QuickSort.java
package praktikum1;

public class QuickSort {
    public static void quickSort(int[] array, int begin, int end) {
        if (begin < end) {
            int pivotIndex = partition(array, begin, end);
            quickSort(array, begin, pivotIndex - 1);
            quickSort(array, pivotIndex + 1, end);
        }
    }

    private static int partition(int[] array, int begin, int end) {
        int pivot = array[end];
        int i = begin - 1;

        for (int j = begin; j < end; j++) {
            if (array[j] >= pivot) { // Bedingung für absteigende Reihenfolge
                i++;
                int temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }
        int temp = array[i + 1];
        array[i + 1] = array[end];
        array[end] = temp;

        return i + 1;
    }

    public static void main(String[] args) {
        int[] array = {99, 0, 33, 27, 87, 80, 65, 77, 39, 88};
        System.out.println("Unsortiertes Array:");
        for (int num : array) System.out.print(num + " ");
        System.out.println();

        quickSort(array, 0, array.length - 1);

        System.out.println("Absteigend sortiertes Array:");
        for (int num : array) System.out.print(num + " ");
    }
}
